SmartDeviceLink/SDLCacheFileManager.m

```diff
@@ -0,0 +1,284 @@
+}
+
+/**
+ * Handles request to update archive file. Will create a new archive file if one is not present.
+ *
+ * @param iconURL The System request URL used to retrieve the icon
+ * @param iconFilePath The directory path where the icon file is held
+ * @param archiveFile Current archive file to update
+*/
+- (BOOL)updateArchiveFileWithIconURL:(NSString *)iconURL
+                        iconFilePath:(NSString *)iconFilePath
+                         archiveFile:(SDLIconArchiveFile *)archiveFile
+                               error:(NSError **)error {
+    NSMutableArray *archiveArray = [archiveFile.lockScreenIconCaches mutableCopy];
+    
+    // Need to remove duplicate SDLLockScreenIconCache object if handling expired icon
+    for (SDLLockScreenIconCache *iconCacheCopy in archiveArray) {
+        if (![iconCacheCopy.iconUrl isEqualToString:iconURL]) { continue; }
+
+        [archiveArray removeObject:iconCacheCopy];
+        break;
+    }
+    
+    // Add the new file to the archive
+    [archiveArray addObject:[[SDLLockScreenIconCache alloc] initWithIconUrl:iconURL iconFilePath:iconFilePath]];
+    archiveFile.lockScreenIconCaches = archiveArray;
+    
+    // HAX: Update this when we are iOS 11.0+
+    // Write the new archive to disk
+    BOOL writeSuccess = [NSKeyedArchiver archiveRootObject:archiveFile toFile:self.archiveFileDirectory];
+    if (!writeSuccess) {
+        if (!*error) {
+            *error = [NSError errorWithDomain:SDLErrorDomainCacheManager code:SDLCacheManagerErrorUpdateArchiveFileFailure userInfo:nil];
+            SDLLogE(@"Error attempting to write archive file to cache: %@", *error);
+        }
+
+        return NO;
+    }
+
+    return YES;
+}
+
+
+#pragma mark - Directory Getters
+
+- (nullable NSString *)cacheFileBaseDirectory {
+    NSString *cacheDirectory = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).firstObject;
+    return [cacheDirectory stringByAppendingPathComponent:@"/sdl/lockScreenIcon/"];
+}
+
+- (nullable NSString *)archiveFileDirectory {
+    return [self.cacheFileBaseDirectory stringByAppendingPathComponent:@"archiveCacheFile"];
+}
+
+#pragma mark - Download Image
+
+/**
+ * Handles request to download icon from request
+ *
+ * @param requestURL The System request URL used to download icon
+ * @param completion The completion handler is called when the icon download succeeds or fails with error
+*/
+- (void)sdl_downloadIconFromRequestURL:(NSString *)requestURL withCompletionHandler:(ImageRetrievalCompletionHandler)completion {
+    [self sdl_sendDataTaskWithURL:[NSURL URLWithString:requestURL] completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+        if (error != nil) {
+            SDLLogW(@"OnSystemRequest (lock screen icon) HTTP download task failed: %@", error);
+            return completion(nil, error);
+        }
+        
+        UIImage *icon = [UIImage imageWithData:data];
+        return completion(icon, nil);
```

Comment:
Check if `icon` is `nil`. If it is, return an error instead because the data couldn't be formed into an image.